src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/AccountValidator.java

```diff
@@ -27,5 +27,7 @@ public AccountValidator(RestRequest request, BytesReference ref, Settings esSett
         super(request, ref, esSettings, param);
         allowedKeys.put("current_password", DataType.STRING);
         mandatoryKeys.add("current_password");
+        mandatoryOrKeys.add("hash");
```

Comment:
Does this mean user always need to pass key "hash": "" even if user just want to pass string password or vice-versa ?

Reply:
No, only hash or password (or any other key present in `mandatoryOrKeys` should be present. I think that value of `mandatoryOrKey` is minimal as parent or child class can add another value to it and the validation won't work anymore. I'd suggest to remove it.

Reply:
That is correct. One of the mandatory keys must be provided (hash or password) and other can be absent. Other APIs like role_mapping do the same. 

I agree that it is not very flexible but works for this use-case. 
I can remove it but code might fail at a much later stage only while comparing for hash and password.

Reply:
I suggest to remove the code and deprecate `mandatoryOrKeys`. It is easy to misuse and break validation accidentally. As there is already code that checks that either "password" or "hash" is present, this code is not necessary.

Reply:
This should probably be a separate issue and propose a better validation structure to use in place of this. 

I do understand that this is not flexible when it comes to `mandatory or keys` in multiple groups eg: (f1 or f2) and (f3 or f4) but at the same time, this is not enforced. 

But before we deprecate and remove this, we should have something that will 
- define the schema of api params instead of digging through the action code.
- fail at early stage (at the beginning of processing request)
- have more rigid bad request validation done here instead of in action code which is more likely to be missed (error-prone). 

I strongly feel we deprecate and remove this only when we have better structure in place as this does it basic job (not saying it does the above perfectly).
 
If you insist I will remove this to unblock the PR but would like to hear your thoughts.

Reply:
I don't suggest to deprecate `mandatoryOrKeys` in this PR. My proposal is to avoid introducing a new dependency on it. As there is only one more `RolesMappingValidator` validator that depends on `mandatoryOrKeys`, there is not much need to make it consistent with other validators.

Later we can decide whether it is worth extending `mandatoryOrKeys` or not. IMO, two cases where it may be useful do not provide enough justification to make it more generic especially that there is still logic in `AccountApiAction` to deal with `password` and `hash`.

I do not insist, but strongly recommend to remove the dependency on `mandatoryOrKeys` from the `AccountValidator`.

Reply:
I absolutely don't mind removing it from the PR but I would like to know the following which is not clear to me:
- why are we deprecating something that works?
- what if the new improved solution is based out of this ? 
- what is the pain point of this in the use cases we have?

Reply:
>why are we deprecating something that works?

While it works, it is fragile and is not flexible.

>what if the new improved solution is based out of this?

It depends on the effort and the value the new improved solution provides. If the current functionality can be easily extended (I doubt), we may decide to extend it instead of deprecating

>what is the pain point of this in the use cases we have

A duplicate check that has to be introduced anyway